ibeckermayer/teleport-interview
# Logout

Adds logout functionality

## backend
Adds a goroutine-safe method `SessionManager.DeleteSession()` for deleting a session from the session manager, and the logout handler that calls it. For edge cases where the "api/logout" request is sent with a missing/improperly-formatted Authorization header, or the token sent in the Authorization header is not found in the `SessionManager`, I've elected to simply log the discrepancy and return a 204 success response to the client so that it still reacts the same as a normal logout (where a session of sessionID is deleted). Open to better ideas/security concerns.

## frontend
The frontend route structure is revamped to direct the user directly to the "/dashboard" route, which is now protected by an `<Authenticated>` wrapper. 

Note: In `<Authenticated>` (and elsewhere), I've elected to use conditional rendering with `<Redirect>` over the `useEffect` hook with `history.push`; unlike the `useEffect` approach, conditional rendering prevents the app from flashing the primary component on the screen before redirecting the user (e.g. see changes starting on lines 25/26 of `Login.jsx`)

## To test
To check manually that the `SessionManager` is behaving as expected, it can be helpful to add `log.Printf("%v", lh.sm)` to line 73 of `login.go` and line 34 of `logout.go`

Run
```bash
docker-compose -f docker-compose-dev.yml up
```
and navigate to https://0.0.0.0:8080

Try navigating directly to each of
* https://0.0.0.0:8080
* https://0.0.0.0:8080/login
* https://0.0.0.0:8080/dashboard

while logged out, and see that they behave as expected. Login with username/pwd = `dev@goteleport.com` and check that you're redirected to the dashboard. 

Try each of the links above again and see that they behave as expected while logged in. 

Hit the "Logout" button and see that you're redirected to /login. Confirm that the `SessionManager` is behaving as it should via the log outputs.
